zookeeper-server/src/main/java/org/apache/zookeeper/server/ReferenceCountedACLCache.java

```diff
@@ -51,7 +51,7 @@
     /**
      * these are the number of acls that we have in the datatree
      */
-    long aclIndex = 0;
+    volatile long aclIndex = 0;
 
```

Comment:
@jhalterman 
Look at the call chain:
```
[1]FileTxnSnapLog#restore--->FileSnap#deserialize--->SerializeUtils.deserializeSnapshot---dt.deserialize--aclCache.deserialize(ia);
[2]Learner#syncWithLeader-->ZKDatabase#deserializeSnapshot[2]--->SerializeUtils.deserializeSnapshot---dt.deserialize--aclCache.deserialize(ia);
```
[1] is for restoring from the snapshot when server start/restart; [2] is for the learner to sync with leader with snap cmd;
the aclIndex is unlikely to be read by multiple threads? Am I right?

Reply:
From what I can tell, the `FileTxnSnapLog#restore` call that happens on startup occurs in the main thread. Later, `Learner#syncWithLeader` occurs in the `QuorumPeer`'s thread.

There appear to be other possible threads as well. Ex:

```FollowerRequestProcessor.run->FileRequestProcessor.processRequest->ZookeeperServer.processTxn->ZkDatabase.processTxn->DataTree.processTxn->DataTree.createNode```

and

```WorkerService.run->CommitProcessor.doWork->FileRequestProcessor.processRequest->ZookeeperServer.processTxn->ZkDatabase.processTxn->DataTree.processTxn->DataTree.createNode```

But overall I'd ask, if `aclIndex` doesn't need to be threadsafe, then why is the rest of `ReferenceCountedACLCache` written to be threadsafe?

Reply:
`ReferenceCountedACLCache.incrementIndex` is only called by `ReferenceCountedACLCache.convertActls` which has been decorated by `synchronized`, so is it threadsafe?

Reply:
I agree with @maoling , aclIndex is guarded by a lock. I don't think volatile is necessary here.

Reply:
The locks just ensure that `aclIndex` is not read/written concurrently, but `volatile` is still needed to ensure that any thread can see the writes made by other threads, otherwise they might be cached. 

http://tutorials.jenkov.com/java-concurrency/volatile.html

Reply:
Yes, a synchronized block also guarantees that the changes to variables inside the protected block is seen by all entering the critical area.
Just double checked also from reference to JCIP :)